### ISSUE TITLE
Unable to Start in Mono - Any Ideas?

### ISSUE THREAD

**Issue by smadam813** [themotleyfool/Klondike#82]
```
mono ./Klondike.SelfHost.exe --interactive --port=8080
Running interactively
2015-05-05 14:28:26,967 [1] INFO  Klondike.SelfHost.KlondikeService - Using ServerFactory Nowin
2015-05-05 14:28:27,203 [1] INFO  NuGet.Lucene.LucenePackageRepository - Refreshing index. Package count: 0

Unhandled Exception:
System.InvalidProgramException: Invalid IL code in (wrapper delegate-invoke) <Module>:invoke_callvirt_CancellationTokenRegistration_CancellationToken&_Action`1<object>_object (System.Threading.CancellationToken&,System.Action`1<object>,object): IL_0057: castclass 0x00000007


  at Microsoft.AspNet.SignalR.Infrastructure.CancellationTokenExtensions.SafeRegister (CancellationToken cancellationToken, System.Action`1 callback, System.Object state) [0x00000] in <filename unknown>:0
  at Microsoft.AspNet.SignalR.Hosting.HostDependencyResolverExtensions.InitializeResolverDispose (IDependencyResolver resolver, CancellationToken hostShutdownToken) [0x00000] in <filename unknown>:0
  at Microsoft.AspNet.SignalR.Hosting.HostDependencyResolverExtensions.InitializeHost (IDependencyResolver resolver, System.String instanceName, CancellationToken hostShutdownToken) [0x00000] in <filename unknown>:0
  at Owin.OwinExtensions.UseSignalRMiddleware[HubDispatcherMiddleware] (IAppBuilder builder, System.Object[] args) [0x00000] in <filename unknown>:0
  at Owin.OwinExtensions.RunSignalR (IAppBuilder builder, Microsoft.AspNet.SignalR.HubConfiguration configuration) [0x00000] in <filename unknown>:0
  at NuGet.Lucene.Web.SignalR.SignalRMapper+<>c__DisplayClass1.<MapSignalR>b__0 (IAppBuilder map) [0x00000] in <filename unknown>:0
  at Owin.MapExtensions.Map (IAppBuilder app, PathString pathMatch, System.Action`1 configuration) [0x00000] in <filename unknown>:0
  at Owin.MapExtensions.Map (IAppBuilder app, System.String pathMatch, System.Action`1 configuration) [0x00000] in <filename unknown>:0
  at NuGet.Lucene.Web.SignalR.SignalRMapper.MapSignalR (IAppBuilder app, Microsoft.AspNet.SignalR.HubConfiguration hubConfiguration) [0x00000] in <filename unknown>:0
  at NuGet.Lucene.Web.Startup.RegisterSignalR (IContainer container, IAppBuilder app) [0x00000] in <filename unknown>:0
  at NuGet.Lucene.Web.Startup.Start (IAppBuilder app, IContainer container) [0x00000] in <filename unknown>:0
  at Klondike.Startup.Start (IAppBuilder app, IContainer container) [0x00000] in <filename unknown>:0
  at Klondike.SelfHost.SelfHostStartup.Start (IAppBuilder app, IContainer container) [0x00000] in <filename unknown>:0
  at NuGet.Lucene.Web.Startup.Configuration (IAppBuilder app) [0x00000] in <filename unknown>:0
  at Microsoft.Owin.Hosting.Engine.HostingEngine.ResolveApp (Microsoft.Owin.Hosting.Engine.StartContext context) [0x00000] in <filename unknown>:0
  at Microsoft.Owin.Hosting.Engine.HostingEngine.Start (Microsoft.Owin.Hosting.Engine.StartContext context) [0x00000] in <filename unknown>:0
  at Microsoft.Owin.Hosting.WebApp.StartImplementation (IServiceProvider services, Microsoft.Owin.Hosting.StartOptions options, System.Action`1 startup) [0x00000] in <filename unknown>:0
  at Microsoft.Owin.Hosting.WebApp.Start (Microsoft.Owin.Hosting.StartOptions options, System.Action`1 startup) [0x00000] in <filename unknown>:0
  at Klondike.SelfHost.KlondikeService.OnStart (System.String[] args) [0x00000] in <filename unknown>:0
  at Klondike.SelfHost.KlondikeService.RunInteractivley () [0x00000] in <filename unknown>:0
  at (wrapper remoting-invoke-with-check) Klondike.SelfHost.KlondikeService:RunInteractivley ()
  at Klondike.SelfHost.Program.Main (System.String[] args) [0x00000] in <filename unknown>:0
[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidProgramException: Invalid IL code in (wrapper delegate-invoke) <Module>:invoke_callvirt_CancellationTokenRegistration_CancellationToken&_Action`1<object>_object (System.Threading.CancellationToken&,System.Action`1<object>,object): IL_0057: castclass 0x00000007


  at Microsoft.AspNet.SignalR.Infrastructure.CancellationTokenExtensions.SafeRegister (CancellationToken cancellationToken, System.Action`1 callback, System.Object state) [0x00000] in <filename unknown>:0
  at Microsoft.AspNet.SignalR.Hosting.HostDependencyResolverExtensions.InitializeResolverDispose (IDependencyResolver resolver, CancellationToken hostShutdownToken) [0x00000] in <filename unknown>:0
  at Microsoft.AspNet.SignalR.Hosting.HostDependencyResolverExtensions.InitializeHost (IDependencyResolver resolver, System.String instanceName, CancellationToken hostShutdownToken) [0x00000] in <filename unknown>:0
  at Owin.OwinExtensions.UseSignalRMiddleware[HubDispatcherMiddleware] (IAppBuilder builder, System.Object[] args) [0x00000] in <filename unknown>:0
  at Owin.OwinExtensions.RunSignalR (IAppBuilder builder, Microsoft.AspNet.SignalR.HubConfiguration configuration) [0x00000] in <filename unknown>:0
  at NuGet.Lucene.Web.SignalR.SignalRMapper+<>c__DisplayClass1.<MapSignalR>b__0 (IAppBuilder map) [0x00000] in <filename unknown>:0
  at Owin.MapExtensions.Map (IAppBuilder app, PathString pathMatch, System.Action`1 configuration) [0x00000] in <filename unknown>:0
  at Owin.MapExtensions.Map (IAppBuilder app, System.String pathMatch, System.Action`1 configuration) [0x00000] in <filename unknown>:0
  at NuGet.Lucene.Web.SignalR.SignalRMapper.MapSignalR (IAppBuilder app, Microsoft.AspNet.SignalR.HubConfiguration hubConfiguration) [0x00000] in <filename unknown>:0
  at NuGet.Lucene.Web.Startup.RegisterSignalR (IContainer container, IAppBuilder app) [0x00000] in <filename unknown>:0
  at NuGet.Lucene.Web.Startup.Start (IAppBuilder app, IContainer container) [0x00000] in <filename unknown>:0
  at Klondike.Startup.Start (IAppBuilder app, IContainer container) [0x00000] in <filename unknown>:0
  at Klondike.SelfHost.SelfHostStartup.Start (IAppBuilder app, IContainer container) [0x00000] in <filename unknown>:0
  at NuGet.Lucene.Web.Startup.Configuration (IAppBuilder app) [0x00000] in <filename unknown>:0
  at Microsoft.Owin.Hosting.Engine.HostingEngine.ResolveApp (Microsoft.Owin.Hosting.Engine.StartContext context) [0x00000] in <filename unknown>:0
  at Microsoft.Owin.Hosting.Engine.HostingEngine.Start (Microsoft.Owin.Hosting.Engine.StartContext context) [0x00000] in <filename unknown>:0
  at Microsoft.Owin.Hosting.WebApp.StartImplementation (IServiceProvider services, Microsoft.Owin.Hosting.StartOptions options, System.Action`1 startup) [0x00000] in <filename unknown>:0
  at Microsoft.Owin.Hosting.WebApp.Start (Microsoft.Owin.Hosting.StartOptions options, System.Action`1 startup) [0x00000] in <filename unknown>:0
  at Klondike.SelfHost.KlondikeService.OnStart (System.String[] args) [0x00000] in <filename unknown>:0
  at Klondike.SelfHost.KlondikeService.RunInteractivley () [0x00000] in <filename unknown>:0
  at (wrapper remoting-invoke-with-check) Klondike.SelfHost.KlondikeService:RunInteractivley ()
  at Klondike.SelfHost.Program.Main (System.String[] args) [0x00000] in <filename unknown>:0
```

**Comment by smadam813** [comment]
Looks like this may be related https://github.com/aspnet/SignalR-Server/issues/93

**Comment by smadam813** [comment]
Works now, you can close at your leisure :-)

**Comment by chriseldredge** [comment]
Looks like using mono 3.12.1 works but 4.0.1 does not.

**Comment by benjamincburns** [comment]
Just to confirm - Installed mono 4.0.1 on latest ubuntu from mono PPA - same issue. Self-compiled/installed mono 3.12.1 and this issue no longer presents.

**Comment by chriseldredge** [comment]
The root cause is [mono 29665](https://bugzilla.xamarin.com/show_bug.cgi?id=29665#c5).

The bug status shows fixed in master, but it looks like it didn't make it into 4.0.1.

**Comment by chriseldredge** [comment]
Since this is a bug in mono I'm closing this as wontfix.

If mono 4.x does not have this resolved when Klondike 2.0 is released, I'll add a disclaimer to the documentation about the problem.

**Comment by chriseldredge** [comment]
Verified this is fixed in mono 4.2.0.
